src/libsystemd/sd-journal/journal-file.c

```diff
@@ -2130,16 +2150,50 @@ static int generic_array_get(
                         return r;
 
                 k = journal_file_entry_array_n_items(o);
-                if (i < k) {
-                        p = le64toh(o->entry_array.items[i]);
-                        goto found;
-                }
+                if (i < k)
+                        break;
 
                 i -= k;
                 t += k;
                 a = le64toh(o->entry_array.next_entry_array_offset);
         }
 
+        /* If we've found the right location, now look for the first non-corrupt entry object (in the right
+         * direction). */
+
+        while (a > 0) {
+                uint64_t k;
+
+                r = journal_file_move_to_object(f, OBJECT_ENTRY_ARRAY, a, &o);
+                if (r < 0)
+                        return r;
+
+                k = journal_file_entry_array_n_items(o);
+                if (k == 0)
+                        break;
+
+                if (i == UINT64_MAX)
+                        i = direction == DIRECTION_DOWN ? 0 : k - 1;
+
+                do {
+                        p = le64toh(o->entry_array.items[i]);
+
+                        r = journal_file_move_to_object(f, OBJECT_ENTRY, p, &e);
```

Comment:
The call of `journal_file_move_to_object()` below is not necessary any more because of the call here?